Winston unit test failed in CI

**Describe the bug**

I got this failed unit test in CI (this is the 1st time I see it): 

```
@chainsafe/lodestar-utils:   112 passing (334ms)
@chainsafe/lodestar-utils:   1 failing
@chainsafe/lodestar-utils:   1) winston logger
@chainsafe/lodestar-utils:        child logger
@chainsafe/lodestar-utils:          Should parse child module:
@chainsafe/lodestar-utils:       AssertionError: expected '' to equal '[A B C]            \u001b[33mwarn\u001b[39m: test'
@chainsafe/lodestar-utils:       + expected - actual
@chainsafe/lodestar-utils:       +[A B C]            warn: test
@chainsafe/lodestar-utils:       
@chainsafe/lodestar-utils:       at Context.<anonymous> (test/unit/logger/winston.test.ts:110:28)
@chainsafe/lodestar-utils: 
@chainsafe/lodestar-utils: error Command failed with exit code 1.
@chainsafe/lodestar-utils: info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
@chainsafe/lodestar-db: $ mocha 'test/unit/**/*.test.ts'
```

**Expected behavior**
All unit tests are passed all the time.

--- a/packages/cli/src/util/logger.ts
+++ b/packages/cli/src/util/logger.ts
@@ -1,4 +1,4 @@
-import {ILogger, LogLevel, TransportType, WinstonLogger} from "@chainsafe/lodestar-utils";
+import {ILogger, LogLevel, TransportType, TransportOpts, WinstonLogger} from "@chainsafe/lodestar-utils";
 
 export function errorLogger(): ILogger {
   return new WinstonLogger({level: LogLevel.error});
@@ -8,8 +8,10 @@ export function errorLogger(): ILogger {
  * Setup a CLI logger, common for beacon, validator and dev commands
  */
 export function getCliLogger(args: {logLevel?: LogLevel; logLevelFile?: LogLevel}, paths: {logFile?: string}): ILogger {
-  return new WinstonLogger({level: args.logLevel}, [
-    {type: TransportType.console},
-    ...(paths.logFile ? [{type: TransportType.file, filename: paths.logFile, level: args.logLevelFile}] : []),
-  ]);
+  const transports: TransportOpts[] = [{type: TransportType.console}];
+  if (paths.logFile) {
+    transports.push({type: TransportType.file, filename: paths.logFile, level: args.logLevelFile});
+  }
+
+  return new WinstonLogger({level: args.logLevel}, transports);
 }

--- a/packages/lodestar/test/utils/logger.ts
+++ b/packages/lodestar/test/utils/logger.ts
@@ -1,5 +1,5 @@
 /* eslint-disable @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-assignment */
-import {WinstonLogger, LogLevel, TransportType} from "@chainsafe/lodestar-utils";
+import {WinstonLogger, LogLevel, TransportType, TransportOpts} from "@chainsafe/lodestar-utils";
 export {LogLevel};
 
 /**
@@ -11,10 +11,12 @@ export {LogLevel};
  * ```
  */
 export function testLogger(module?: string, defaultLogLevel = LogLevel.error, logFile?: string): WinstonLogger {
-  return new WinstonLogger({level: getLogLevelFromEnvs() || defaultLogLevel, module}, [
-    {type: TransportType.console},
-    ...(logFile ? [{type: TransportType.file, filename: logFile, level: LogLevel.debug}] : []),
-  ]);
+  const transports: TransportOpts[] = [{type: TransportType.console}];
+  if (logFile) {
+    transports.push({type: TransportType.file, filename: logFile, level: LogLevel.debug});
+  }
+
+  return new WinstonLogger({level: getLogLevelFromEnvs() || defaultLogLevel, module}, transports);
 }
 
 function getLogLevelFromEnvs(): LogLevel | null {

--- a/packages/utils/src/logger/interface.ts
+++ b/packages/utils/src/logger/interface.ts
@@ -13,6 +13,15 @@ export enum LogLevel {
   silly = "silly",
 }
 
+export const logLevelNum: {[K in LogLevel]: number} = {
+  [LogLevel.error]: 0,
+  [LogLevel.warn]: 1,
+  [LogLevel.info]: 2,
+  [LogLevel.verbose]: 3,
+  [LogLevel.debug]: 4,
+  [LogLevel.silly]: 5,
+};
+
 // eslint-disable-next-line @typescript-eslint/naming-convention
 export const LogLevels = Object.values(LogLevel);
 

--- a/packages/utils/src/logger/transport.ts
+++ b/packages/utils/src/logger/transport.ts
@@ -5,11 +5,13 @@ import TransportStream from "winston-transport";
 export enum TransportType {
   console = "console",
   file = "file",
+  stream = "stream",
 }
 
 export type TransportOpts =
   | {type: TransportType.console; level?: LogLevel}
-  | {type: TransportType.file; level?: LogLevel; filename: string};
+  | {type: TransportType.file; level?: LogLevel; filename: string}
+  | {type: TransportType.stream; level?: LogLevel; stream: NodeJS.WritableStream};
 
 export function fromTransportOpts(transportOpts: TransportOpts): TransportStream {
   switch (transportOpts.type) {
@@ -24,6 +26,13 @@ export function fromTransportOpts(transportOpts: TransportOpts): TransportStream
       return new transports.File({
         level: transportOpts.level,
         filename: transportOpts.filename,
+        handleExceptions: true,
+      });
+
+    case TransportType.stream:
+      return new transports.Stream({
+        level: transportOpts.level,
+        stream: transportOpts.stream,
         handleExceptions: true,
       });
   }

--- a/packages/utils/src/logger/winston.ts
+++ b/packages/utils/src/logger/winston.ts
@@ -3,7 +3,7 @@
  */
 
 import {createLogger, Logger} from "winston";
-import {Context, defaultLogLevel, ILogger, ILoggerOptions, LogLevel} from "./interface";
+import {Context, defaultLogLevel, ILogger, ILoggerOptions, LogLevel, logLevelNum} from "./interface";
 import chalk from "chalk";
 import {getFormat} from "./format";
 import {Writable} from "stream";
@@ -22,14 +22,21 @@ export class WinstonLogger implements ILogger {
   private _transportOptsArr: TransportOpts[];
 
   constructor(options: Partial<ILoggerOptions> = {}, transportOptsArr: TransportOpts[] = [defaultTransportOpts]) {
+    // `options.level` can override the level in the transport
+    // This is necessary for child logger opts to take effect
+    let minLevel = options?.level || defaultLogLevel;
+    for (const transportOpts of transportOptsArr) {
+      transportOpts.level = minLevel = getMinLevel([minLevel, transportOpts.level || defaultLogLevel]);
+    }
+
     this.winston = createLogger({
       level: options?.level || defaultLogLevel,
       defaultMeta: {module: options?.module || ""} as DefaultMeta,
       format: getFormat(options || {}),
-      transports: transportOptsArr.map(fromTransportOpts),
+      transports: transportOptsArr.map((transportOpts) => fromTransportOpts(transportOpts)),
       exitOnError: false,
     });
-    this._level = this.getMinLevel(transportOptsArr.map((opts) => opts.level || defaultLogLevel));
+    this._level = minLevel;
     // Store for child logger
     this._options = options;
     this._transportOptsArr = transportOptsArr;
@@ -79,18 +86,18 @@ export class WinstonLogger implements ILogger {
 
   private createLogEntry(level: LogLevel, message: string, context?: Context, error?: Error): void {
     // don't propagate if silenced or message level is more detailed than logger level
-    if (this.winston.levels[level] > this.winston.levels[this._level]) {
+    if (logLevelNum[level] > logLevelNum[this._level]) {
       return;
     }
     this.winston[level](message, {context, error});
   }
+}
 
-  /** Return the min LogLevel from multiple transports */
-  private getMinLevel(levels: LogLevel[]): LogLevel {
-    return levels.reduce(
-      // error: 0, warn: 1, info: 2, ...
-      (minLevel, level) => (this.winston.levels[level] > this.winston.levels[minLevel] ? level : minLevel),
-      defaultLogLevel
-    );
-  }
+/** Return the min LogLevel from multiple transports */
+function getMinLevel(levels: LogLevel[]): LogLevel {
+  return levels.reduce(
+    // error: 0, warn: 1, info: 2, ...
+    (minLevel, level) => (logLevelNum[level] > logLevelNum[minLevel] ? level : minLevel),
+    defaultLogLevel
+  );
 }

--- a/packages/utils/test/unit/logger/winston.test.ts
+++ b/packages/utils/test/unit/logger/winston.test.ts
@@ -1,16 +1,26 @@
 import fs from "fs";
 import path from "path";
 import rimraf from "rimraf";
+import {Writable} from "stream";
 import {Context, ILogger, LodestarError, LogFormat, logFormats, WinstonLogger} from "../../../src";
 import {expect} from "chai";
 import {TransportType} from "../../../src/logger/transport";
 
-describe("winston logger", () => {
-  const tmpDir = fs.mkdtempSync("lodestar-winston-test");
-  after(() => {
-    rimraf.sync(tmpDir);
-  });
+/**
+ * To capture Winston output in memory
+ */
+class WritableMemory extends Writable {
+  chunks: Buffer[] = [];
+  _write(chunk: Buffer): void {
+    this.chunks.push(chunk);
+  }
 
+  getAsString(): string {
+    return Buffer.concat(this.chunks).toString("utf8");
+  }
+}
+
+describe("winston logger", () => {
   describe("winston logger format and options", () => {
     interface ITestCase {
       id: string;
@@ -63,15 +73,11 @@ describe("winston logger", () => {
       const {id, message, context, error, output} = typeof testCase === "function" ? testCase() : testCase;
       for (const format of logFormats) {
         it(`${id} ${format} output`, async () => {
-          const filename = path.join(tmpDir, `${id.replace(/\s+/g, "-")}-${format}.txt`);
-
-          const logger = new WinstonLogger({format, hideTimestamp: true}, [{type: TransportType.file, filename}]);
+          const stream = new WritableMemory();
+          const logger = new WinstonLogger({format, hideTimestamp: true}, [{type: TransportType.stream, stream}]);
           logger.warn(message, context, error);
 
-          // Allow the file transport to persist the file
-          await new Promise((r) => setTimeout(r, 10));
-
-          const allOutput = fs.readFileSync(filename, "utf8").trim();
+          const allOutput = stream.getAsString().trim();
           expect(allOutput).to.equal(output[format]);
         });
       }
@@ -96,9 +102,8 @@ describe("winston logger", () => {
 
   describe("child logger", () => {
     it("Should parse child module", async () => {
-      const filename = path.join(tmpDir, "child-logger-test.txt");
-
-      const logger = new WinstonLogger({hideTimestamp: true, module: "A"}, [{type: TransportType.file, filename}]);
+      const stream = new WritableMemory();
+      const logger = new WinstonLogger({hideTimestamp: true, module: "A"}, [{type: TransportType.stream, stream}]);
       const childB = logger.child({module: "B"});
       const childC = childB.child({module: "C"});
       childC.warn("test");
@@ -106,8 +111,32 @@ describe("winston logger", () => {
       // Allow the file transport to persist the file
       await new Promise((r) => setTimeout(r, 10));
 
-      const allOutput = fs.readFileSync(filename, "utf8").trim();
+      const allOutput = stream.getAsString().trim();
       expect(allOutput).to.equal("[A B C]            \u001b[33mwarn\u001b[39m: test");
+    });
+  });
+
+  describe("Log to file", () => {
+    const tmpDir = fs.mkdtempSync("lodestar-winston-test");
+
+    it("Should log to file", async () => {
+      const filename = path.join(tmpDir, "child-logger-test.txt");
+
+      const logger = new WinstonLogger({hideTimestamp: true, module: "A"}, [{type: TransportType.file, filename}]);
+      logger.warn("test");
+
+      // Wait for file to exist
+      for (let i = 0; i < 200; i++) {
+        await new Promise((r) => setTimeout(r, 10));
+        if (fs.existsSync(filename)) break;
+      }
+
+      const allOutput = fs.readFileSync(filename, "utf8").trim();
+      expect(allOutput).to.equal("[A]                \u001b[33mwarn\u001b[39m: test");
+    });
+
+    after(() => {
+      rimraf.sync(tmpDir);
     });
   });
 });

--- a/packages/utils/test/unit/logger/winston.test.ts
+++ b/packages/utils/test/unit/logger/winston.test.ts
@@ -2,8 +2,8 @@ import fs from "fs";
 import path from "path";
 import rimraf from "rimraf";
 import {Writable} from "stream";
-import {Context, ILogger, LodestarError, LogFormat, logFormats, WinstonLogger} from "../../../src";
 import {expect} from "chai";
+import {Context, LodestarError, LogFormat, logFormats, LogLevel, sleep, WinstonLogger} from "../../../src";
 import {TransportType} from "../../../src/logger/transport";
 
 /**
@@ -77,26 +77,25 @@ describe("winston logger", () => {
           const logger = new WinstonLogger({format, hideTimestamp: true}, [{type: TransportType.stream, stream}]);
           logger.warn(message, context, error);
 
-          const allOutput = stream.getAsString().trim();
-          expect(allOutput).to.equal(output[format]);
+          expect(stream.getAsString().trim()).to.equal(output[format]);
         });
       }
     }
   });
 
   describe("profile", () => {
-    it("should log profile", () => {
-      const logger: ILogger = new WinstonLogger();
+    it("should log profile", async () => {
+      const stream = new WritableMemory();
+      const logger = new WinstonLogger({hideTimestamp: true, module: "A"}, [{type: TransportType.stream, stream}]);
 
       logger.profile("test");
+      await sleep(10);
+      // Stop profile of 'test'. Logging will now take place:
+      // '2021-03-31 20:47:09 []                 info: test  - duration=10ms'
+      logger.profile("test");
 
-      setTimeout(function () {
-        //
-        // Stop profile of 'test'. Logging will now take place:
-        //   '17 Jan 21:00:00 - info: test duration=10ms'
-        //
-        logger.profile("test");
-      }, 10);
+      // Do not include the time since it can vary
+      expect(stream.getAsString().trim()).to.include("\u001b[32minfo\u001b[39m: test  - duration=");
     });
   });
 
@@ -108,16 +107,30 @@ describe("winston logger", () => {
       const childC = childB.child({module: "C"});
       childC.warn("test");
 
-      // Allow the file transport to persist the file
-      await new Promise((r) => setTimeout(r, 10));
+      expect(stream.getAsString().trim()).to.equal("[A B C]            \u001b[33mwarn\u001b[39m: test");
+    });
 
-      const allOutput = stream.getAsString().trim();
-      expect(allOutput).to.equal("[A B C]            \u001b[33mwarn\u001b[39m: test");
+    it("Should log to child at a lower logLevel", () => {
+      const stream = new WritableMemory();
+      const logger = new WinstonLogger({hideTimestamp: true, module: "A"}, [
+        {type: TransportType.stream, stream, level: LogLevel.info},
+      ]);
+
+      const childB = logger.child({module: "B", level: LogLevel.debug});
+
+      logger.debug("Should not be logged");
+      childB.debug("Should be logged");
+
+      expect(stream.getAsString().trim()).to.equal("[A B]             \u001b[34mdebug\u001b[39m: Should be logged");
     });
   });
 
   describe("Log to file", () => {
-    const tmpDir = fs.mkdtempSync("lodestar-winston-test");
+    let tmpDir: string;
+
+    before(() => {
+      tmpDir = fs.mkdtempSync("test-lodestar-winston-test");
+    });
 
     it("Should log to file", async () => {
       const filename = path.join(tmpDir, "child-logger-test.txt");
@@ -131,8 +144,8 @@ describe("winston logger", () => {
         if (fs.existsSync(filename)) break;
       }
 
-      const allOutput = fs.readFileSync(filename, "utf8").trim();
-      expect(allOutput).to.equal("[A]                \u001b[33mwarn\u001b[39m: test");
+      const output = fs.readFileSync(filename, "utf8").trim();
+      expect(output).to.equal("[A]                \u001b[33mwarn\u001b[39m: test");
     });
 
     after(() => {

